File: Tests/GlasstualTests/TLOInternetAddressLookupTests.m
/* *********************************************************************
 * Copyright (c) 2026 Codeux Software, LLC & respective contributors.
 * Please see Acknowledgements.pdf for additional information.
 *********************************************************************** */

#import <XCTest/XCTest.h>

#import "TLOInternetAddressLookupPrivate.h"

NS_ASSUME_NONNULL_BEGIN

@interface TLOInternetAddressLookupDelegateSpy : NSObject <TLOInternetAddressLookupDelegate>
@end

@implementation TLOInternetAddressLookupDelegateSpy

- (void)internetAddressLookupReturnedAddress:(NSString *)address
{
}

- (void)internetAddressLookupFailed
{
}

@end

@interface TLOInternetAddressLookupTests : XCTestCase
@end

@implementation TLOInternetAddressLookupTests

- (NSHTTPURLResponse *)responseWithStatusCode:(NSInteger)statusCode
{
	return [[NSHTTPURLResponse alloc] initWithURL:[NSURL URLWithString:@"https://example.invalid/address"]
									   statusCode:statusCode
									  HTTPVersion:@"HTTP/1.1"
									 headerFields:nil];
}

- (void)testAddressParserTrimsAndAcceptsEnabledIPv4
{
	NSData *data = [@"  203.0.113.42\n" dataUsingEncoding:NSUTF8StringEncoding];

	NSString *address = [TLOInternetAddressLookup addressFromData:data
														 response:[self responseWithStatusCode:200]
														allowIPv4:YES
														allowIPv6:NO];

	XCTAssertEqualObjects(address, @"203.0.113.42");
}

- (void)testAddressParserAcceptsEnabledIPv6
{
	NSData *data = [@"2001:db8::1" dataUsingEncoding:NSUTF8StringEncoding];

	NSString *address = [TLOInternetAddressLookup addressFromData:data
														 response:[self responseWithStatusCode:200]
														allowIPv4:NO
														allowIPv6:YES];

	XCTAssertEqualObjects(address, @"2001:db8::1");
}

- (void)testAddressParserRejectsDisabledOrMalformedAddresses
{
	NSHTTPURLResponse *response = [self responseWithStatusCode:200];
	NSData *IPv4Data = [@"203.0.113.42" dataUsingEncoding:NSUTF8StringEncoding];
	NSData *invalidData = [@"not an address" dataUsingEncoding:NSUTF8StringEncoding];

	XCTAssertNil([TLOInternetAddressLookup addressFromData:IPv4Data response:response allowIPv4:NO allowIPv6:YES]);
	XCTAssertNil([TLOInternetAddressLookup addressFromData:invalidData response:response allowIPv4:YES allowIPv6:YES]);
}

- (void)testAddressParserRejectsBadStatusAndOversizedResponse
{
	NSData *addressData = [@"203.0.113.42" dataUsingEncoding:NSUTF8StringEncoding];
	NSData *oversizedData = [NSMutableData dataWithLength:1025];

	XCTAssertNil([TLOInternetAddressLookup addressFromData:addressData
												  response:[self responseWithStatusCode:500]
												 allowIPv4:YES
												 allowIPv6:YES]);
	XCTAssertNil([TLOInternetAddressLookup addressFromData:oversizedData
												  response:[self responseWithStatusCode:200]
												 allowIPv4:YES
												 allowIPv6:YES]);
}

- (void)testLookupDefaultsToBothAddressFamilies
{
	TLOInternetAddressLookupDelegateSpy *delegate = [TLOInternetAddressLookupDelegateSpy new];
	TLOInternetAddressLookup *lookup = [[TLOInternetAddressLookup alloc] initWithDelegate:delegate];

	XCTAssertTrue(lookup.IPv4AddressIsValid);
	XCTAssertTrue(lookup.IPv6AddressIsValid);
}

@end

NS_ASSUME_NONNULL_END

File: Tests/GlasstualTests/TLOInternetAddressLookupTests.swift
/*  *********************************************************************
 * Copyright (c) 2026 Codeux Software, LLC & respective contributors.
 * Please see Acknowledgements.pdf for additional information.
 *********************************************************************** */

import Foundation
@testable import Glasstual
import Testing

@MainActor
private final class TLOInternetAddressLookupDelegateSpy: NSObject, InternetAddressLookupDelegate {
	func internetAddressLookupReturnedAddress(_: String) {}

	func internetAddressLookupFailed() {}
}

@MainActor
@Suite("Internet address lookup")
struct TLOInternetAddressLookupTests {
	private func response(statusCode: Int) throws -> HTTPURLResponse {
		let url = try #require(URL(string: "https://example.invalid/address"))

		return try #require(
			HTTPURLResponse(
				url: url,
				statusCode: statusCode,
				httpVersion: "HTTP/1.1",
				headerFields: nil
			)
		)
	}

	@Test("An IPv4 answer is trimmed of the whitespace the service pads it with")
	func addressParserTrimsAndAcceptsEnabledIPv4() throws {
		let data = Data("  203.0.113.42\n".utf8)
		let address = try InternetAddressLookup.address(
			from: data,
			response: response(statusCode: 200),
			allowIPv4: true,
			allowIPv6: false
		)

		#expect(address == "203.0.113.42")
	}

	@Test("An IPv6 answer is accepted when IPv6 is the enabled family")
	func addressParserAcceptsEnabledIPv6() throws {
		let data = Data("2001:db8::1".utf8)
		let address = try InternetAddressLookup.address(
			from: data,
			response: response(statusCode: 200),
			allowIPv4: false,
			allowIPv6: true
		)

		#expect(address == "2001:db8::1")
	}

	@Test("An address of a disabled family, or no address at all, is rejected")
	func addressParserRejectsDisabledOrMalformedAddresses() throws {
		let successfulResponse = try response(statusCode: 200)
		let IPv4Data = Data("203.0.113.42".utf8)
		let invalidData = Data("not an address".utf8)

		#expect(InternetAddressLookup.address(
			from: IPv4Data,
			response: successfulResponse,
			allowIPv4: false,
			allowIPv6: true
		) == nil)
		#expect(InternetAddressLookup.address(
			from: invalidData,
			response: successfulResponse,
			allowIPv4: true,
			allowIPv6: true
		) == nil)
	}

	@Test("A failed status, or a body larger than an address could be, is rejected")
	func addressParserRejectsBadStatusAndOversizedResponse() throws {
		let addressData = Data("203.0.113.42".utf8)
		let oversizedData = Data(count: 1025)
		let failedResponse = try response(statusCode: 500)
		let successfulResponse = try response(statusCode: 200)

		#expect(InternetAddressLookup.address(
			from: addressData,
			response: failedResponse,
			allowIPv4: true,
			allowIPv6: true
		) == nil)
		#expect(InternetAddressLookup.address(
			from: oversizedData,
			response: successfulResponse,
			allowIPv4: true,
			allowIPv6: true
		) == nil)
	}

	/// The lookup and its delegate are main-actor isolated; the parser above
	/// is not.
	@Test("A lookup asks for both address families until it is told otherwise")
	func lookupDefaultsToBothAddressFamilies() {
		let delegate = TLOInternetAddressLookupDelegateSpy()
		let lookup = InternetAddressLookup(delegate: delegate)

		#expect(lookup.ipv4AddressIsValid)
		#expect(lookup.ipv6AddressIsValid)
	}
}
